SQL: JDBC: fix temporary directory locked test errors in Windows

bpintea opened the pull request:
The tests involving a temporary directory (containing the JDBC JAR) fail
on Windows because they can't be deleted, due to still being in use.
This PR forces an early closing of the JAR file, which mitigates
the failure by giving the JVM more time to collect any open FDs.
(Calling the System.gc() in the tests is another working alternative
fix.)

The stream-based JAR access is taken care by disabling the cache usage
(which has the same effect of closing the JAR).

Fixes #56882.

bpintea committed:
Fix temp dir locked errors

The tests involving a temporary directory (containing the JDBC JAR) fail
on Windows because they can't be deleted, due to still being in use.
This commit forces a premature closing of the JAR file, which mitigates
the failure by giving the JVM more time to collect any open FDs.
(Calling the System.gc() in the tests is another working alternative
fix.)

The stream-based JAR access is taken care by disabling the cache usage

bpintea committed:
Fix style errors

- supress JarFile usage warning
- remove unused import

bpintea committed:
Address review comments

Wrap closing the JAR into a try-block, ignore any failure caused by
it: we can continue for as long as we've got the Manifest

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/SQL)

matriv commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/ClientVersion.java:
```diff
@@ -87,7 +88,11 @@ static Manifest getManifest(URL url) throws IOException {
+                    Manifest manifest = jarConn.getManifest(); // in case of a fat JAR, this would return the outermost JAR's manifest
+                    try {
+                        jarConn.getJarFile().close(); // prevent locked file errors in Windows. The Manifest has been read by now
+                    } catch (IOException e) {/* ignore any closing failure */}
```
Maybe we should log here, since this is not only for testing.

bpintea replied:
We could, but where: the CLI client uses `java.util.logging`, but JDBC uses it's own logging. Would hooking into either just for this warning - since it's not really an error - be justified?

matriv replied:
My bad, sorry, misread that the return is within the catch block.

matriv replied:
Probably not I guess...

matriv commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/ClientVersion.java:
```diff
@@ -87,7 +88,11 @@ static Manifest getManifest(URL url) throws IOException {
+                    try {
+                        jarConn.getJarFile().close(); // prevent locked file errors in Windows. The Manifest has been read by now
+                    } catch (IOException e) {/* ignore any closing failure */}
+                    return manifest;
```
Shouldn't this be in a `finally` block to return not only in case of exception?

bpintea replied:
Not sure I get your suggestion: the function returns irrespective of an `IOException` being raised. Or are you rather suggesting to return a version (in the calling function), also when an `IOExceptoin` is raised (which we currently propagate into a `IllegalArgumentException`)?

matriv replied:
My bad, sorry, misread that the return is within the catch block.

matriv reviewed (commented):
Left a couple of comments.

astefan commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/ClientVersion.java:
```diff
@@ -87,7 +88,11 @@ static Manifest getManifest(URL url) throws IOException {
-                    return jarConn.getManifest(); // in case of a fat JAR, this would return the outermost JAR's manifest
+                    Manifest manifest = jarConn.getManifest(); // in case of a fat JAR, this would return the outermost JAR's manifest
+                    try {
+                        jarConn.getJarFile().close(); // prevent locked file errors in Windows. The Manifest has been read by now
```
Here both `getJarFile()` and `.close()` can throw an IOException. `getManifest()` call above does use `getJarFile()` so the danger of an IOException being thrown while calling `getJarFile()` wouldn't be ignored. But, would make sense to call `jarConn.getJarFile()` and reuse this object in both `getManifest()` and `.close()` calls?

bpintea replied:
> why is this .close() needed now, but wasn't needed before?

We just didn't test with JAR non-entry URLs. The tests were using file (i.e. setting the `ClientVersion.CURRENT`) and entry JAR URLs (i.e. the JarInJar test).

> Here both getJarFile() and .close() can throw an IOException. 

True. And while the former's exception should surface, the latter (if it can ever occur) is non-critical.

> would make sense to call jarConn.getJarFile() and reuse this object in both getManifest() and .close() calls?

That exposes a bit more of "deprecated" surface and is more verbose, but I see how this would make sense. I've applied your suggestion.

astefan reviewed (approved):
LGTM, but left one comment.
There is another aspect that is still unknown to me: why is this `.close()` needed now, but wasn't needed before?

bpintea committed:
Address review comments

Extract Jar file out of the connection and work directly with it when
extracting the manifeset and closing it.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/fix_locked_jar_file_tests

costin commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/ClientVersion.java:
```diff
@@ -87,7 +89,12 @@ static Manifest getManifest(URL url) throws IOException {
                 JarURLConnection jarConn = (JarURLConnection) conn;
                 if (jarConn.getEntryName() == null) { // the URL points to a JAR file
-                    return jarConn.getManifest(); // in case of a fat JAR, this would return the outermost JAR's manifest
+                    JarFile jar = jarConn.getJarFile();
```
I suggest using the try/close approach - it's similar but more concise:

```suggestion
if (jarConn.getEntryName() == null) {
 try (JarFile jar = jarconn.getJarFile()) {
      return jar.getManifest();
  }
}
```

bpintea replied:
Beautiful, forgot `JarFile` implements `AutoCloseable`. Thanks!

bpintea replied:
For the record, that will still interrupt the loading if closing fails (while it shouldn't), but OTOH, I can't imagine a case where opening would work, but closing not.

costin reviewed (approved):
LGTM

matriv reviewed (approved):
LGTM

bpintea committed:
Address review comments

- make use of JarFile AutoCloseable if'ace (try-with-resources)
